File: 2392-successful-pairs-of-spells-and-potions/successful-pairs-of-spells-and-potions.js
var successfulPairs = function(spells, potions, success) {
    let output = []
    potions.sort((a, b)=> a - b)

    function binarySearch(spell) {
        let i = 0
        let j = potions.length - 1

        while (i <= j) {
            let mid = Math.floor((i + j) / 2)
            if (potions[mid] * spell >= success) {
                j = mid - 1
            }
            else i = mid + 1
        }
        return i
    }

    for (let i = 0; i < spells.length; i++){
        let spell = spells[i]
        let index = binarySearch(spell)
        output[i] = potions.length - index
    }

    return output
};

File: 2392-successful-pairs-of-spells-and-potions/successful-pairs-of-spells-and-potions.py
class Solution:
    def successfulPairs(self, spells: List[int], potions: List[int], success: int) -> List[int]:
        output = [None] * len(spells)
        potions.sort()

        def bs(spell):
            i = 0
            j = len(potions) - 1

            while (i <= j):
                mid = (i + j) // 2

                if (potions[mid] * spell >= success):
                    j = mid - 1
                else:
                    i = mid + 1

            return i
        
        for i in range(len(spells)):
            spell = spells[i]
            output[i] = len(potions) - bs(spell)

        return output
